controllers/state_of_the_world.go

```diff
@@ -0,0 +1,132 @@
+			kuadrantv1beta2.AuthPolicyKind,
+			kuadrantv1beta2.RateLimitPolicyKind,
+		),
+		controller.WithObjectKinds(
+			kuadrantv1beta1.KuadrantKind,
+			schema.GroupKind{Group: corev1.GroupName, Kind: "ConfigMap"}),
+		controller.WithObjectLinks(
+			kuadrantv1beta1.LinkKuadrantToGatewayClasses,
+			kuadrantv1beta1.LinkKuadrantToTopologyConfigMap,
+		),
+		controller.WithReconcile(buildReconciler(client)),
+	}
+
+	return controller.NewController(controllerOpts...)
+}
+
+func buildWatcher[T controller.Object](obj T, resource schema.GroupVersionResource, namespace string, options ...controller.RunnableBuilderOption[T]) controller.RunnableBuilder {
+	return controller.Watch(obj, resource, namespace, options...)
+}
+
+func buildReconciler(client *dynamic.DynamicClient) controller.ReconcileFunc {
+	topologyFileReconciler := TopologyFileReconciler{Client: client}
+
+	reconciler := &controller.Workflow{
+		Precondition: func(ctx context.Context, resourceEvents []controller.ResourceEvent, _ *machinery.Topology) {
+			logger := controller.LoggerFromContext(ctx).WithName("event logger")
+			for _, event := range resourceEvents {
+				// log the event
+				obj := event.OldObject
+				if obj == nil {
+					obj = event.NewObject
+				}
+				values := []any{
+					"type", event.EventType.String(),
+					"kind", obj.GetObjectKind().GroupVersionKind().Kind,
+					"namespace", obj.GetNamespace(),
+					"name", obj.GetName(),
+				}
+				if event.EventType == controller.UpdateEvent && logger.V(1).Enabled() {
+					values = append(values, "diff", cmp.Diff(event.OldObject, event.NewObject))
+				}
+				logger.Info("new event", values...)
+			}
+		},
+		Tasks: []controller.ReconcileFunc{
+			topologyFileReconciler.Reconcile,
+		},
+	}
+
+	return reconciler.Run
+}
+
+type TopologyFileReconciler struct {
+	Client *dynamic.DynamicClient
+}
+
+func (r *TopologyFileReconciler) Reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology) {
+	logger := controller.LoggerFromContext(ctx).WithName("topology file")
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "topology",
+			Namespace: kuadrantNamespace,
+		},
+		Data: map[string]string{
+			"topology": topology.ToDot(),
+		},
+	}
+
+	configMapRes := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "configmaps"}
+	unstructuredCM, _ := controller.Destruct(cm)
+
+	targets := topology.Objects().Items(func(object machinery.Object) bool {
+		if object.GetName() == cm.GetName() && object.GetNamespace() == cm.GetNamespace() {
+			return true
+		}
+		return false
+	})
+
+	if len(targets) == 0 {
+		_, err := r.Client.Resource(configMapRes).Namespace(cm.Namespace).Create(context.TODO(), unstructuredCM, metav1.CreateOptions{})
+		if err != nil {
+			logger.Error(err, "failed to write topology configmap")
+		}
+		return
+	}
+	_, err := r.Client.Resource(configMapRes).Namespace(cm.Namespace).Update(context.TODO(), unstructuredCM, metav1.UpdateOptions{})
```

Comment:
Let's use the passed in context in the client calls rather than a new context
```suggestion
	logger := controller.LoggerFromContext(ctx).WithName("topology file")
	cm := &corev1.ConfigMap{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "topology",
			Namespace: kuadrantNamespace,
		},
		Data: map[string]string{
			"topology": topology.ToDot(),
		},
	}

	configMapRes := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "configmaps"}
	unstructuredCM, _ := controller.Destruct(cm)

	targets := topology.Objects().Items(func(object machinery.Object) bool {
		if object.GetName() == cm.GetName() && object.GetNamespace() == cm.GetNamespace() {
			return true
		}
		return false
	})

	if len(targets) == 0 {
		_, err := r.Client.Resource(configMapRes).Namespace(cm.Namespace).Create(ctx, unstructuredCM, metav1.CreateOptions{})
		if err != nil {
			logger.Error(err, "failed to write topology configmap")
		}
		return
	}
	_, err := r.Client.Resource(configMapRes).Namespace(cm.Namespace).Update(ctx, unstructuredCM, metav1.UpdateOptions{})
```